test/bats/test.bats

```diff
@@ -172,13 +173,8 @@ teardown() {
 }
 
 @test "required labels audit test" {
-  wait_for_process $WAIT_TIME $SLEEP_TIME "kubectl get k8srequiredlabels.constraints.gatekeeper.sh ns-must-have-gk -o json | jq '.status.violations[]'"
-
-  violations=$(kubectl get k8srequiredlabels.constraints.gatekeeper.sh ns-must-have-gk -o json | jq '.status.violations | length')
-  [[ "$violations" -eq 6 ]]
-
-  totalViolations=$(kubectl get k8srequiredlabels.constraints.gatekeeper.sh ns-must-have-gk -o json | jq '.status.totalViolations')
-  [[ "$totalViolations" -eq 6 ]]
+  run wait_for_process $WAIT_TIME $SLEEP_TIME "check_audit_violations 6"
```

Comment:
I wonder if we could keep the body of the test in this file, without needing to jump to helpers to see the implementation (which, in this case, contains the real test body).

One way would be expanding the argument of wait_for_process to contain the comparison, the other would be to have the check_audit_violations right before the test.

As a side note, is the run + assert really needed here? I am seeing mixed use in the tests file, and here the success depends only on wait_for_process to succeed.

Reply:
> keep the body of the test in this file

+1

Reply:
run + assert is probably not necessary, I think any non-zero exit code will work. The majority of the code uses it, so I mainly was just trying to keep things uniform.

Probably the simplest way to keep things in the same spot is to move the function definition to the bats file.

Trying to do an assertion across multiple lines of code and avoiding flaking due to eventual consistency means that running a retry loop around the actual comparison function is probably the best way.

Remember, for this test two things need to be correct: the violation count and the actual list of reported violations.

Reply:
This guide suggestions using run to assert success is superfluous:

https://github.com/bats-core/bats-core#testing

I removed the function to see what it'd look like. I'm not sure it makes the code more readable vs. defining the test in a function above the unit test, though it does remove the ambiguity as to whether `set -e` is in effect.

Reply:
I saw that the flakiness was still there (apparently the old namespace was still hanging around).

I switched most tests to use configmaps rather than namespaces, which should be faster and require less load on the cluster.

I also cleaned up the e2e tests generally (removing redundant waits, proper enforcement checks, unnecessary `run` invocations, etc.)

Reply:
Trying to just check the audit count without the use of functions makes it impossible to debug why the test fails due to lack of messaging.

Moving the interior audit test function into the bats file to get the requested readability. Bash does not really do nested function definitions.

Reply:
@ritazh @fedepaol ready for another look

I ran this version of the e2e tests on GitHub a number of times, and could not get them to fail